Refactor missing namespace version

rly opened the pull request:
Based on [commentary](https://github.com/hdmf-dev/hdmf/commit/d432709374c7e8435f3a47fe29f906495d71ae97#r37363245) from @yarikoptic and @ajtritt, refactor, and:
1. expose the magic value that represents a missing version: None
2. write "None" as the version when writing the spec to disk
3. set the `SpecNamespace.version` to None if "None" is read as the version from disk
4. set the `SpecNamespace.version` to None if "unversioned" is read as the version from disk
5. make sure the "unversioned" and "None" version namespaces are not loaded when a versioned namespace is present
6. add a bunch of tests to cover these edge-cases and backward-compatibility support

rly committed:
Change missing namespace version back to None

rly committed:
Revert "Change missing namespace version back to None"

This reverts commit ed601d393391821cdb6fcc25cdf49c44041640a6.

rly committed:
Refactor magic string "unversioned" to class variable

rly committed:
Fix tests

rly committed:
Make getter return the UNVERSIONED string if version is somehow removed

rly committed:
Remove redundant "unversioned" on write

rly committed:
Fix last commit

yarikoptic commented on src/hdmf/spec/namespace.py:
```diff
@@ -83,13 +85,16 @@ def contact(self):
+        version = self.get('version', None)
+        if version is None:
+            self['version'] = SpecNamespace.UNVERSIONED
+        return self.get('version')
```
`return self.get('version', None) or SpecNamespace.UNVERSIONED` probably do the same if version can't be numeric ;-) could also benefit of it is an empty list or an empty string for some reason

rly replied:
The code as written also sets self['version']. How about changing line 95 to:
```python
if not version:
```

yarikoptic replied:
I love functions with side effects, they are so much fun to debug ;-)
If anything, could also just not use version variable - used only once

rly replied:
True, side effects are not great. I'll go with your suggestion.

rly committed:
Simplify getter, remove side effect

rly committed:
Fix flake8

rly committed:
Merge branch 'dev' into fix/ns_version2

ajtritt committed:
make version None when no namespace version supplied

ajtritt commented on src/hdmf/spec/namespace.py:
```diff
@@ -83,13 +85,16 @@ def contact(self):
+        String, list, or tuple with the version or SpecNamespace.UNVERSIONED
+        if the version is missing or empty
+        """
+        return self.get('version', None) or SpecNamespace.UNVERSIONED
```
This would return a bool. Did you mean 

`return self.get('verzion', SpecNamespace.UNVERSIONED)`

ajtritt commented on src/hdmf/spec/namespace.py:
```diff
@@ -83,13 +85,16 @@ def contact(self):
+        String, list, or tuple with the version or SpecNamespace.UNVERSIONED
+        if the version is missing or empty
+        """
+        return self.get('version', None) or SpecNamespace.UNVERSIONED
```
This would return a bool. Did you mean this?

`return self.get('version', SpecNamespace.UNVERSIONED)`

rly replied:
Interestingly, this does not return a bool. 
```python                            
>>> 'a' or 'b'
'a'
>>> False or 'b'
'b'
>>> None or 'b'
'b'
>>> None or None
>>>                      
```
As written, if `version` is somehow '', [], or (), then `SpecNamespace.UNVERSIONED` will be returned. 

If `self.get('version', SpecNamespace.UNVERSIONED)` is used, then if `version` is somehow '', [], or (), then those would get returned.

ajtritt commented on src/hdmf/backends/hdf5/h5tools.py:
```diff
@@ -280,10 +280,7 @@ def write(self, **kwargs):
-                    group_name = '%s/unversioned' % ns_name
-                else:
-                    group_name = '%s/%s' % (ns_name, namespace.version)
+                group_name = '%s/%s' % (ns_name, namespace.version)
```
I changed the default to be `None`. This would make the group name "None". I don't have a problem with that, since I don't think it would impact reading the namespace

yarikoptic replied:
cool.  When it is loaded from a file, would `"None"` string be casted to a proper `None` for the `.version`? Just checking, didn't spot anything obvious in the diff

rly replied:
This is now resolved in `SpecNamespace.__init__`

rly replied:
@yarikoptic Thanks for checking that!

ajtritt reviewed (commented):
@rly I made a couple of changes and comments. Please look them over... I suspect @yarikoptic would give a thumbs up.

rly committed:
Fix test assertion

rly committed:
Set NS version to UNVERSIONED value if init receives str(UNVERSIONED)

The HDF5 backend will serialize the UNVERSIONED value None to "None". On 
read, init needs to detect the string "None" and set version to None 
accordingly.

rly committed:
Fix unversioned being used before versioned if multiple ext versions

rly committed:
Remove extraneous assertion in test which was breaking in py35

rly committed:
Merge branch 'dev' into fix/ns_version2

rly committed:
Merge branch 'dev' into fix/ns_version2
